### PR TITLE
CATROID-556 Edit variables

### PR DESCRIPTION
**Pull request by khushal1707**
https://jira.catrob.at/browse/CATROID-556
Content (Value) of Variable is editable from data list, option will appear on long-press menu.

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -2069,6 +2069,6 @@ needs read and write access to it. You can always change permissions through you
     <string name="hertz_symbol">㎐</string>
 
     <string name="sans_serif_font" translatable="false">sans_serif</string>
-    <string name="serif_font" translatable="false">serif</string>
+    <string name="serif_font" translatable="false">serif</string><string name="edit">Edit</string><string name="data_value">Value</string>
```
Comment (wslany) [inline catroid/src/main/res/values/strings.xml:2072]:
```suggestion
    <string name="serif_font" translatable="false">serif</string>
    <string name="edit">Edit</string>
    <string name="data_value">Value</string>
```

This should fix most of the lint warnings of https://jenkins.catrob.at/job/Catroid/job/PR-3925/1/analysis/

**Review by wslany — CHANGES_REQUESTED** [review]
Awesome! Very nice, thanks a lot!!

Please add an espresso test, then it will be ready I think.

File: catroid/src/main/res/values/strings.xml (hunk 1 above)
Comment (khushal1707) [reply catroid/src/main/res/values/strings.xml:2072]: Working on the changes

**Comment by khushal1707** [comment]
Hi @wslany sir, I have added espresso test and some other minor requested changes, please review it and also suggest any further improvement or changes required in this PR. In the meantime, can I work on new issues?

**Review by wslany — CHANGES_REQUESTED** [review]
Thank you, and sorry for the late reply!!!!

Please fix the lint warnings as seen in https://jenkins.catrob.at/job/Catroid/job/PR-3925/11/analysis/

**Comment by wslany** [comment]
Thanks! Could you just please change the years in the files you edited in the header to the new year 2021?

**Comment by khushal1707** [comment]
> Thanks! Could you just please change the years in the files you edited in the header to the new year 2021?
> 
> Like in
> Copyright (C) 2010-2018
> =>
> Copyright (C) 2010-2021

Done!

**Review by wslany — APPROVED** [review]

**Comment by devsebastian** [comment]
Hi @khushal1707  seems like the values of the data variables change back to the original values (ie. 0) every time you press the play button. Please look into this issue! everything else seems to be working fine!

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/DataListFragment.kt [hunk 2]
```diff
@@ -293,6 +292,34 @@ class DataListFragment : Fragment(),
...
     }
 
+    private fun showEditDialog(selectedItems: List<UserData<*>>) {
+        val item = selectedItems[0]
+        val builder = TextInputDialog.Builder(requireContext())
+
+        builder.setHint(getString(R.string.data_value))
+            .setText(item.value.toString())
+            .setPositiveButton(
+                getString(R.string.ok)
```
Comment (devsebastian) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/DataListFragment.kt:302]: also please change the button title from "ok" to "save" as described in the ticket.

**Review by devsebastian — CHANGES_REQUESTED** [review]

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/DataListFragment.kt (hunk 2 above)
Comment (khushal1707) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/DataListFragment.kt:302]: Sure!

**Comment by khushal1707** [comment]
> Hi @khushal1707 seems like the values of the data variables change back to the original values (ie. 0) every time you press the play button. Please look into this issue! everything else seems to be working fine!

Do we need to retain that data and use it in subsequent runs Or just retain before the new data get stored in variable?
I think all variable gets reset to initial value when we re-run it, because there is no option of resume after exiting from the stage.
Retaining of variable's value can be implemented in Resume, i.e. on resuming it will retain new value(that we enter using edit) or the previously stored values. Correct me if I am wrong, Thanks.

**Comment by wslany** [comment]
> > Hi @khushal1707 seems like the values of the data variables change back to the original values (ie. 0) every time you press the play button. Please look into this issue! everything else seems to be working fine!
> 
> @wslany @devsebastian Do we need to retain that data and use it in subsequent runs Or just retain before the new data get stored in variable?
> I think all variable gets reset to initial value when we re-run it, because there is no option of resume after exiting from the stage.
> Retaining of variable's value can be implemented in Resume, i.e. on resuming it will retain new value(that we enter using edit) or the previously stored values. Correct me if I am wrong, Thanks.

@khushal1707 you are right, the variable values are reset. Later (but not in this ticket) we will add a "resume" feature, then it will make more sense. But we can use it already now to inspect and copy/paste into other apps very large text content.

**Comment by devsebastian** [comment]
Ok then @khushal1707 please just rename the ok button, everything else works fine, I'll then push the ticket to PO review

**Comment by khushal1707** [comment]
> > > Hi @khushal1707 seems like the values of the data variables change back to the original values (ie. 0) every time you press the play button. Please look into this issue! everything else seems to be working fine!
> > 
> > 
> > @wslany @devsebastian Do we need to retain that data and use it in subsequent runs Or just retain before the new data get stored in variable?
> > I think all variable gets reset to initial value when we re-run it, because there is no option of resume after exiting from the stage.
> > Retaining of variable's value can be implemented in Resume, i.e. on resuming it will retain new value(that we enter using edit) or the previously stored values. Correct me if I am wrong, Thanks.
> 
> @khushal1707 you are right, the variable values are reset. Later (but not in this ticket) we will add a "resume" feature, then it will make more sense. But we can use the edit function already now to inspect and copy/paste into other apps very large text content.

Inspecting and copy/pasting large text content can be done by implement code. So, I think there is no need of change in logical part?

**Comment by khushal1707** [comment]
> Ok then @khushal1707 please just rename the ok button, everything else works fine, I'll then push the ticket to PO review

Sure!

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/utils/UserDataItemRVInteractionWrapper.java [hunk 3]
```diff
@@ -68,6 +68,17 @@ public void performRename(String newName) {
...
 	}
 
+	public void performEdit(String newValue) {
+		onChildView(R.id.title_view)
+				.perform(longClick());
+		onView(withText(R.string.edit))
+				.perform(click());
+		onView(withId(R.id.input_edit_text))
+				.perform(replaceText(newValue), closeSoftKeyboard());
+		onView(withText(R.string.ok))
```
Comment (devsebastian) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/utils/UserDataItemRVInteractionWrapper.java:78]: I think this too needs to be changed to R.string.save!

Comment (devsebastian) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/utils/UserDataItemRVInteractionWrapper.java:78]: Please change R.string.ok to R.string.save, as changed in the Edit Dialog.

**Review by devsebastian — CHANGES_REQUESTED** [review]

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/utils/UserDataItemRVInteractionWrapper.java (hunk 3 above)
Comment (khushal1707) [reply catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/formulaeditor/utils/UserDataItemRVInteractionWrapper.java:78]: Oh sorry, doing it.

**Review by devsebastian — CHANGES_REQUESTED** [review]
Please don't include commits/files that aren't required by the ticket. (like deletion of .idea/.name file) Please revert this unwanted change before the commit could be merged.😅

**Comment by khushal1707** [comment]
> Please don't include commits/files that aren't required by the ticket. (like deletion of .idea/.name file) Please revert this unwanted change before the commit could be merged.😅

Could you please tell me where is this file located (in which folder)?

**Comment by khushal1707** [comment]
> Please don't include commits/files that aren't required by the ticket. (like deletion of .idea/.name file) Please revert this unwanted change before the commit could be merged.😅

Sorry for the trouble, Its done!

**Review by devsebastian — APPROVED** [review]

**Comment by wslany** [comment]
Thanks!

**Pull request merged**